dash/include/dash/algorithm/Copy.h

```diff
@@ -105,108 +130,83 @@ template <
-      size_type max_elem_per_unit = pattern.local_size(local_pos.unit);
-      // Local offset of first element in input range at current unit:
-      auto l_in_first_idx  = local_pos.index;
-      // Maximum number of elements to copy from current unit:
-      auto num_unit_elem   = max_elem_per_unit - l_in_first_idx;
-      // Number of elements left to copy:
-      auto total_elem_left = num_elem_total - num_elem_copied;
-      // Number of elements to copy in this iteration.
-      auto num_copy_elem   = num_unit_elem;
-      if (num_copy_elem > total_elem_left) {
-        num_copy_elem = total_elem_left;
+
+  ContiguousRangeSet<GlobInputIt> range_set{begin, end};
+
+  std::vector<local_copy_chunk<nonconst_value_type>> local_chunks;
+
+  //
+  // Copy elements from every unit:
+  //
+
+  for (auto range : range_set) {
+
+    auto cur_in        = range.first;
+    auto num_copy_elem = range.second;
+
+    DASH_ASSERT_GT(num_copy_elem, 0, "Number of elements to copy is 0");
+    auto dest_ptr = out_first + num_elem_copied;
+
+    // handle local data locally
+    if (cur_in.is_local()) {
+      // if the chunk is less than a page or if it is the only transfer
+      // don't bother post-poning it
+      auto src_ptr = cur_in.local();
+      if (num_elem_total == num_copy_elem ||
+          DASH__ARCH__PAGE_SIZE > num_copy_elem*sizeof(ValueType)) {
+        std::copy(src_ptr, src_ptr + num_copy_elem, dest_ptr);
+      } else {
+        // larger chunks are handled later to allow overlap
+        local_copy_chunk<nonconst_value_type> chunk{src_ptr, dest_ptr,
+                                                    num_copy_elem};
+        local_chunks.push_back(chunk);
       }
-      DASH_ASSERT_GT(num_copy_elem, 0,
-                     "Number of element to copy is 0");
-      DASH_LOG_TRACE("dash::copy_impl",
-                     "start g_idx:",    cur_in_first.pos(),
-                     "->",
-                     "unit:",           local_pos.unit,
-                     "l_idx:",          l_in_first_idx,
-                     "->",
-                     "unit elements:",  num_unit_elem,
-                     "max elem/unit:",  max_elem_per_unit,
-                     "get elements:",   num_copy_elem,
-                     "total:",          num_elem_total,
-                     "copied:",         num_elem_copied,
-                     "left:",           total_elem_left);
-      auto dest_ptr = out_first + num_elem_copied;
-      auto src_gptr = cur_in_first.dart_gptr();
-      dart_handle_t handle;
-      dash::internal::get_handle(src_gptr, dest_ptr, num_copy_elem, &handle);
-      num_elem_copied += num_copy_elem;
-      if (handle != DART_HANDLE_NULL) {
-        handles.push_back(handle);
+    } else {
+      auto src_gptr = cur_in.dart_gptr();
+
+      DASH_LOG_TRACE("dash::copy_impl", "dest_ptr", dest_ptr,
+                    "src_gptr", src_gptr, "num_copy_elem", num_copy_elem);
+
+      if (handles != nullptr) {
```

Comment:
Could we move the `#if DASH_COPY_USE_HANDLES` here, instead of this runtime condition. That would make it transparent to the caller, whether handles are used or not.

Reply:
Nope, `copy_impl` is used by `async_copy` too, which needs to used handles no matter what. The compiler will propagate the `nullptr` constant to optimize the runtime check away. `copy_impl` should not be called by the user directly anyway :)